### PR TITLE
docs: migrate to scriv to manage changelog

### PR DESCRIPTION
**Pull request by regisb**
Changelog management was starting to be a hassle:

- there were conflicts every time a PR was merged
- there were conflicts every time we merged the nightly branch in the new release branch, or vice versa.

Now, all changelog entries are stored as separate files in changelog.d, including nightly. Nightly entries will be collected for every major release.

### CODE REVIEW

**Comment by regisb** [comment]
And of course I'm tagging @nedbat here ;)

**Comment by regisb** [comment]
@overhangio/tutor-maintainers FYI I suggest that we migrate to [scriv](https://github.com/nedbat/scriv) for managing changelogs. I expect that we will start maintaining changelogs in plugins as well very soon (I will be working on a PR if we agree on this).

**Comment by ghassanmas** [comment]
Sounds good to me, following this the scriv document here is how I imagine the follow would be:

- Create a new branch to resolve a bug, add new feature...etc 
- Make changes to code files as needed
- **Before commiting** run `scriv create`  
- Edit the new created file accordingly 
- Commit the changes togather: 'Code + Changelog' 


 In case it's a big change i.e. there are multiple commits in the log, then is it expected to create a dedicated commit for the cahngelog?. 
 
Lastly what is mapping between git commit message, and the changelog. Do we give low priority for commit message description that we have changelog, do we use duplicate content, summary...etc

File: changelog.d/scriv/new_fragment.md.j2 [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+Create a changelog entry for every new user-facing change. Please respect the following instructions:
+- Indicate breaking changes by prepending an explosion 💥 character.
+- Prefix your changes with either [Bugfix], [Improvement], [Feature], [Security], [Deprecation].
+- You may optionally append "(by @<author>)" at the end of the line, where "<author>" is either one (just one)
+  of your GitHub username, real name or affiliated organization. These affiliations will be displayed in
+  the release notes for every release.
+-->
+
+<!-- - 💥[Feature] Foobarize the blorginator. This breaks plugins by renaming the `FOO_DO` filter to `BAR_DO`. (by @regisb) -->
+<!-- - [Improvement] This is a non-breaking change. Life is good. (by @billgates) -->
```
Comment (arbrandes) [inline changelog.d/scriv/new_fragment.md.j2:11]: "by @billgates" 😆

File: docs/tutorials/nightly.rst [hunk 2]
```diff
@@ -61,4 +61,4 @@ Making changes to Tutor Nightly
 In general pull requests should be open on the "master" branch of Tutor: the "master" branch is automatically merged on the "nightly" branch at every commit, such that changes made to Tutor releases find their way to Tutor Nightly as soon as they are merged. However, sometimes you want to make changes to Tutor Nightly exclusively, and not to the Tutor releases. This might be the case for instance when upgrading the running version of a third-party service (for instance: Elasticsearch, MySQL), or when the master branch requires specific changes. In that case, you should follow the instructions from the :ref:`contributing` section of the docs, with the following differences:
 
 - Open your pull request on top of the "nightly" branch instead of "master".
-- Add a description of your changes to CHANGELOG-nightly.md instead of CHANGELOG.md
+- Add a description of your changes to by creating a changelog entry with `make changelog-entry`, as in the master branch.
```
Comment (arbrandes) [inline docs/tutorials/nightly.rst:64]:
```suggestion
- Add a description of your changes by creating a changelog entry with `make changelog-entry`, as in the master branch.
```

File: docs/tutor.rst [hunk 3]
```diff
@@ -69,9 +69,9 @@ Releasing a new version
 ~~~~~~~~~~~~~~~~~~~~~~~
 
 - Bump the ``__version__`` value in ``tutor/__about__.py``. (see :ref:`versioning` below)
-- Replace "Unreleased" with the version name and date in CHANGELOG.md.
+- Collect changelog entries with ``make changelog``.
 - Create a commit with the version changelog.
-- Run ``make release``: this will push to the default repo/branch for the current branch.
+- Run ``make release``: this will create a tag and push to the default repo/branch for the current branch.
```
Comment (arbrandes) [inline docs/tutor.rst:74]: Just checking: you're not actually changing the `make release` behavior, right?  Just documenting what it does.

**Review by arbrandes — APPROVED** [review]
One suggestion, one reaction, and one question.  Approved. 👍🏼

File: docs/tutor.rst (hunk 3 above)
Comment (regisb) [reply docs/tutor.rst:74]: Right, exactly.

**Comment by regisb** [comment]
> In case it's a big change i.e. there are multiple commits in the log, then is it expected to create a dedicated commit for the changelog?.
> Lastly what is mapping between git commit message, and the changelog. Do we give low priority for commit message description that we have changelog, do we use duplicate content, summary...etc

The roles of commits/changelog entries will not change after this PR. I have always considered commits and changelog entries to be two _very different things_. Commits are read by developers. Changelogs are read by Tutor users. Commit titles are meant to be short and sweet, while changelog entries should explain, for instance, how to handle breaking changes. Commit texts should describe the thought process that went into fixing issues, while changelogs should focus on the result for end users.

So to answer your question @ghassanmas:

> In case it's a big change i.e. there are multiple commits in the log, then is it expected to create a dedicated commit for the changelog?

There is no general answer to that. It really depends.

> Lastly what is mapping between git commit message, and the changelog. Do we give low priority for commit message description that we have changelog, do we use duplicate content, summary...etc

There is no mapping at all. If some information must be duplicated, then so be it. We must not expect that people who read changelogs also read commit messages. The reverse is true, in most cases. Again, I don't have an answer that will apply to all cases. In general, I'd say: write as many words as possible. It's more work, but it's almost always best for the readers, whether they are developers or users.

**Review by nedbat — APPROVED** [review]

**Comment by ghassanmas** [comment]
Thank you @regisb for the detailed explanation, that exacly what I was looking for. And I do agree with your conclusion  👍🏾

**Comment by kdmccormick** [comment]
I didn't have a chance to review, but this looks great. If using scriv with Tutor goes well, this will help us sell the idea of using it in Open edX too.

**Pull request merged**